new-build doesn't honor v1.6 cabal-spec build-dep semantics

In `Distribution.Backpack.ConfiguredComponent` there's

```hs
newPackageDepsBehaviourMinVersion :: Version
newPackageDepsBehaviourMinVersion = mkVersion [1,7,1]


-- In older cabal versions, there was only one set of package dependencies for
-- the whole package. In this version, we can have separate dependencies per
-- target, but we only enable this behaviour if the minimum cabal version
-- specified is >= a certain minimum. Otherwise, for compatibility we use the
-- old behaviour.
newPackageDepsBehaviour :: PackageDescription -> Bool
newPackageDepsBehaviour pkg =
   specVersion pkg >= newPackageDepsBehaviourMinVersion
```

However, `operational-0.2.3.4` relies on this semantics and is currently broken under `new-build`:

```
$ cabal get operational-0.2.3.4 ; cd operational-0.2.3.4/
Unpacking to operational-0.2.3.4/
$ cabal new-build
Resolving dependencies...
In order, the following will be built (use -v for more details):
 - operational-0.2.3.4 {operational-0.2.3.4-inplace} (lib) (first run)
 - operational-0.2.3.4 {operational-0.2.3.4-inplace-operational-TicTacToe} (exe:operational-TicTacToe) (first run)
Configuring component lib from operational-0.2.3.4
Configuring component exe:operational-TicTacToe from operational-0.2.3.4
Preprocessing library operational-0.2.3.4...
Preprocessing executable 'operational-TicTacToe' for operational-0.2.3.4...
[1 of 1] Compiling Control.Monad.Operational ( src/Control/Monad/Operational.hs, /tmp/operational-0.2.3.4/dist-newstyle/build/x86_64-linux/ghc-8.0.1/operational-0.2.3.4/build/Control/Monad/Operational.o )
[1 of 2] Compiling Control.Monad.Operational ( src/Control/Monad/Operational.hs, /tmp/operational-0.2.3.4/dist-newstyle/build/x86_64-linux/ghc-8.0.1/operational-0.2.3.4/c/operational-TicTacToe/build/operational-TicTacToe/operational-TicTacToe-tmp/Control/Monad/Operational.o )

src/Control/Monad/Operational.hs:24:1: error:
    Failed to load interface for ‘Control.Monad.Identity’
    It is a member of the hidden package ‘mtl-2.2.1’.
    Perhaps you need to add ‘mtl’ to the build-depends in your .cabal file.
    Use -v to see a list of the files searched for.

src/Control/Monad/Operational.hs:25:1: error:
    Failed to load interface for ‘Control.Monad.Trans’
    It is a member of the hidden package ‘mtl-2.2.1’.
    Perhaps you need to add ‘mtl’ to the build-depends in your .cabal file.
    Use -v to see a list of the files searched for.

src/Control/Monad/Operational.hs:32:1: error:
    Failed to load interface for ‘Control.Monad.Reader.Class’
    It is a member of the hidden package ‘mtl-2.2.1’.
    Perhaps you need to add ‘mtl’ to the build-depends in your .cabal file.
    Use -v to see a list of the files searched for.

src/Control/Monad/Operational.hs:33:1: error:
    Failed to load interface for ‘Control.Monad.State.Class’
    It is a member of the hidden package ‘mtl-2.2.1’.
    Perhaps you need to add ‘mtl’ to the build-depends in your .cabal file.
    Use -v to see a list of the files searched for.

src/Control/Monad/Operational.hs:26:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
cabal: Failed to build exe:operational-TicTacToe from operational-0.2.3.4.
```

/cc @ezyang

--- a/cabal-install/Distribution/Client/ProjectBuilding.hs
+++ b/cabal-install/Distribution/Client/ProjectBuilding.hs
@@ -872,7 +872,7 @@ buildAndInstallUnpackedPackage verbosity
 
     let dispname = case elabPkgOrComp pkg of
             ElabPackage _ -> display pkgid
-                ++ " (all, due to Custom setup)"
+                ++ " (all, legacy fallback)"
             ElabComponent comp -> display pkgid
                 ++ " (" ++ maybe "custom" display (compComponentName comp) ++ ")"
 

--- a/cabal-install/Distribution/Client/ProjectPlanning.hs
+++ b/cabal-install/Distribution/Client/ProjectPlanning.hs
@@ -1142,6 +1142,11 @@ elaborateInstallPlan verbosity platform compiler compilerprogdb pkgConfigDB
             -- type, and teach all of the code paths how to handle it.
             -- Once you've implemented this, swap it for the code below.
             = fromMaybe PD.Custom (PD.buildType (elabPkgDescription elab0)) /= PD.Custom
+            -- cabal-format versions prior to 1.8 have different build-depends semantics
+            -- for now it's easier to just fallback to legacy-mode when specVersion < 1.8
+            -- see, https://github.com/haskell/cabal/issues/4121
+              && PD.specVersion pd >= mkVersion [1,8]
+
             {-
             -- Only non-Custom or sufficiently recent Custom
             -- scripts can be build per-component.

